packages/mui-material/src/FilledInput/FilledInput.js

```diff
@@ -70,88 +70,114 @@ const FilledInputRoot = styled(InputBaseRoot, {
-          duration: theme.transitions.duration.shorter,
-          easing: theme.transitions.easing.easeOut,
-        }),
-        pointerEvents: 'none', // Transparent to the hover style.
-      },
-      [`&.${filledInputClasses.focused}:after`]: {
-        // translateX(0) is a workaround for Safari transform scale bug
-        // See https://github.com/mui/material-ui/issues/31766
-        transform: 'scaleX(1) translateX(0)',
+    variants: [
+      {
+        props: { startAdornment: true },
+        style: {
+          paddingLeft: 12,
+        },
       },
-      [`&.${filledInputClasses.error}`]: {
-        '&::before, &::after': {
-          borderBottomColor: (theme.vars || theme).palette.error.main,
+      {
+        props: { endAdornment: true },
+        style: {
+          paddingRight: 12,
         },
       },
-      '&::before': {
-        borderBottom: `1px solid ${
-          theme.vars
-            ? `rgba(${theme.vars.palette.common.onBackgroundChannel} / ${theme.vars.opacity.inputUnderline})`
-            : bottomLineColor
-        }`,
-        left: 0,
-        bottom: 0,
-        // Doing the other way around crash on IE11 "''" https://github.com/cssinjs/jss/issues/242
-        content: '"\\00a0"',
-        position: 'absolute',
-        right: 0,
-        transition: theme.transitions.create('border-bottom-color', {
-          duration: theme.transitions.duration.shorter,
-        }),
-        pointerEvents: 'none', // Transparent to the hover style.
+      {
+        props: { multiline: true },
+        style: {
+          padding: '25px 12px 8px',
+        },
       },
-      [`&:hover:not(.${filledInputClasses.disabled}, .${filledInputClasses.error}):before`]: {
-        borderBottom: `1px solid ${(theme.vars || theme).palette.text.primary}`,
+      {
+        props: { multiline: true, size: 'small' },
+        style: {
+          paddingTop: 21,
+          paddingBottom: 4,
+        },
       },
-      [`&.${filledInputClasses.disabled}:before`]: {
-        borderBottomStyle: 'dotted',
+      {
+        props: { multiline: true, hiddenLabel: true },
+        style: {
+          paddingTop: 16,
+          paddingBottom: 17,
+        },
       },
-    }),
-    ...(ownerState.startAdornment && {
-      paddingLeft: 12,
-    }),
-    ...(ownerState.endAdornment && {
-      paddingRight: 12,
-    }),
-    ...(ownerState.multiline && {
-      padding: '25px 12px 8px',
-      ...(ownerState.size === 'small' && {
-        paddingTop: 21,
-        paddingBottom: 4,
-      }),
-      ...(ownerState.hiddenLabel && {
-        paddingTop: 16,
-        paddingBottom: 17,
-      }),
-      ...(ownerState.hiddenLabel &&
-        ownerState.size === 'small' && {
+      {
+        props: { multiline: true, hiddenLabel: true, size: 'small' },
+        style: {
           paddingTop: 8,
           paddingBottom: 9,
-        }),
-    }),
+        },
+      },
+    ],
+
+    ...Object.entries(theme.palette)
```

Comment:
I think this part should be fixed. If you do this, it means that the styles of `'&::after': {…}, [`&.${filledInputClasses.focused}:after`]: {…}, …` will be multiplied by the amount of palettes.

I should be like this:

```js
variants: [
  …,
  { props: { disableUnderline: false },
    style: {
      '&::after': {
        left: 0,
        bottom: 0,
        // Doing the other way around crash on IE11 "''" https://github.com/cssinjs/jss/issues/242
        content: '""',
        position: 'absolute',
        right: 0,
        transform: 'scaleX(0)',
        transition: theme.transitions.create('transform', {
          duration: theme.transitions.duration.shorter,
          easing: theme.transitions.easing.easeOut,
        }),
        pointerEvents: 'none', // Transparent to the hover style.
      },
      [`&.${filledInputClasses.focused}:after`]: {
        // translateX(0) is a workaround for Safari transform scale bug
        // See https://github.com/mui/material-ui/issues/31766
        transform: 'scaleX(1) translateX(0)',
      },
      [`&.${filledInputClasses.error}`]: {
        '&::before, &::after': {
          borderBottomColor: (theme.vars || theme).palette.error.main,
        },
      },
      '&::before': {
        borderBottom: `1px solid ${
          theme.vars
            ? `rgba(${theme.vars.palette.common.onBackgroundChannel} / ${theme.vars.opacity.inputUnderline})`
            : bottomLineColor
        }`,
        left: 0,
        bottom: 0,
        // Doing the other way around crash on IE11 "''" https://github.com/cssinjs/jss/issues/242
        content: '"\\00a0"',
        position: 'absolute',
        right: 0,
        transition: theme.transitions.create('border-bottom-color', {
          duration: theme.transitions.duration.shorter,
        }),
        pointerEvents: 'none', // Transparent to the hover style.
      },
      [`&:hover:not(.${filledInputClasses.disabled}, .${filledInputClasses.error}):before`]: {
        borderBottom: `1px solid ${(theme.vars || theme).palette.text.primary}`,
      },
      [`&.${filledInputClasses.disabled}:before`]: {
        borderBottomStyle: 'dotted',
      },
    },
  },
  …Object.entries(theme.palette)
    .filter(([, value]) => value.main)
    .map(([color]) => ({
      props: { disableUnderline: false, color },
      style: {
        '&::after': {
            borderBottom: `2px solid ${(theme.vars || theme).palette[color]?.main}`,
         }
      }
    })
]
```